### DISCUSSION TITLE
Disconnects in few seconds after connecting to wifi network

### DISCUSSION THREAD

**Discussion by zeebak** [arendst/Tasmota#11244] [Q&A] [answered]
I have compiled bin based on latest firmware using Gitpod.
I have made only a handful of changes viz.,
the Project name

MODULE    USER_MODULE

FALLBACK_MODULE     USER_MODULE

USER_TEMPLATE    "{\" .......}"

WIFI_CONFIG_TOOL       WIFI_MANAGER

Home assistant auto discovery    true

These are all the changes I made and I think it shouldn't affect the connectivity/Wifi settings. 

The device starts and i enter my wifi ssid in the wifi_manager. It restarts, connects and disconnects after few seconds again starting the Wifi Manager.

Please help, compiling for esp8266

**Comment by barbudor** [comment, marked as answer]
You have intentionally specified to use WifiManager, so behave as expected
Use WIFI_WAIT

**Reply by zeebak** [reply]
Same Problem with WIFI_WAIT.

I thought the WifiManager would only come up if it fails to connect. But my connection was successful. It's confusing, why would Wifi_Config_Tool be retrying if its already connected.

**Reply by zeebak** [reply]
What I am trying to achieve is, When/If wifi fails to connect, I want Tasmota to start in AP mode so that I can enter/re-enter my (changed) Wifi credentials. How do I do this?

**Reply by barbudor** [reply]
So WIFI_MANAGER is the right choice but what you see should occur only the 1st time
Let the device go to WifiManager mode, then it should revert to your SSID after 3 minutes
This should never occur again after, unless your SSID is down

**Reply by zeebak** [reply]
**This is  my_user_config.h  file**

// -- Project -------------------------------------
#define PROJECT                "tasmota"         // PROJECT is used as the default topic delimiter

// If not selected the default will be SONOFF_BASIC
#define MODULE                 USER_MODULE      // [Module] Select default module from tasmota_template.h
#ifdef ESP8266
#define FALLBACK_MODULE        USER_MODULE      // [Module2] Select default module on fast reboot where USER_MODULE is user template
#define USER_TEMPLATE "{\"NAME\":\"JTS\",\"GPIO\":[0,32,0,256,257,258,0,0,33,259,34,35,0,0],\"FLAG\":0,\"BASE\":18}"
#endif  // ESP8266
#ifdef ESP32
#define FALLBACK_MODULE        WEMOS             // [Module2] Select default module on fast reboot where USER_MODULE is user template
//#define USER_TEMPLATE "{\"NAME\":\"ESP32-DevKit\",\"GPIO\":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],\"FLAG\":0,\"BASE\":1}"  // [Template] Set JSON template
#endif  // ESP32

#define SAVE_DATA              1                 // [SaveData] Save changed parameters to Flash (0 = disable, 1 - 3600 seconds)
#define SAVE_STATE             true              // [SetOption0] Save changed power state to Flash (false = disable, true = enable)
#define BOOT_LOOP_OFFSET       1                 // [SetOption36] Number of boot loops before starting restoring defaults (0 = disable, 1..200 = boot loops offset)

// -- Wifi ----------------------------------------
#define WIFI_IP_ADDRESS        "0.0.0.0"         // [IpAddress1] Set to 0.0.0.0 for using DHCP or enter a static IP address
#define WIFI_GATEWAY           "192.168.1.1"     // [IpAddress2] If not using DHCP set Gateway IP address
#define WIFI_SUBNETMASK        "255.255.255.0"   // [IpAddress3] If not using DHCP set Network mask
#define WIFI_DNS               "192.168.1.1"     // [IpAddress4] If not using DHCP set DNS IP address (might be equal to WIFI_GATEWAY)

#define STA_SSID1              ""                // [Ssid1] Wifi SSID
#define STA_PASS1              ""                // [Password1] Wifi password
#define STA_SSID2              ""                // [Ssid2] Optional alternate AP Wifi SSID
#define STA_PASS2              ""                // [Password2] Optional alternate AP Wifi password
#define WIFI_AP_PASSPHRASE     ""                // AccessPoint passphrase. For WPA2 min 8 char, for open use "" (max 63 char).
#define WIFI_CONFIG_TOOL       WIFI_MANAGER        // [WifiConfig] Default tool if wifi fails to connect (default option: 4 - WIFI_RETRY)
                                                 // (WIFI_RESTART, WIFI_MANAGER, WIFI_RETRY, WIFI_WAIT, WIFI_SERIAL, WIFI_MANAGER_RESET_ONLY)
                                                 // The configuration can be changed after first setup using WifiConfig 0, 2, 4, 5, 6 and 7.
#define WIFI_CONFIG_NO_SSID    WIFI_MANAGER      // Default tool if wifi fails to connect and no SSID is configured


// Change ARP interval to 0 to reset arp int.

#define WIFI_ARP_INTERVAL      0                 // [SetOption41] Send gratuitous ARP interval
#define WIFI_SCAN_AT_RESTART   false             // [SetOption56] Scan wifi network at restart for configured AP's
#define WIFI_SCAN_REGULARLY    true              // [SetOption57] Scan wifi network every 44 minutes for configured AP's

// -- Syslog --------------------------------------
#define SYS_LOG_HOST           ""                // [LogHost] (Linux) syslog host
#define SYS_LOG_PORT           514               // [LogPort] default syslog UDP port
#define SYS_LOG_LEVEL          LOG_LEVEL_NONE    // [SysLog] (LOG_LEVEL_NONE, LOG_LEVEL_ERROR, LOG_LEVEL_INFO, LOG_LEVEL_DEBUG, LOG_LEVEL_DEBUG_MORE)
#define SERIAL_LOG_LEVEL       LOG_LEVEL_INFO    // [SerialLog] (LOG_LEVEL_NONE, LOG_LEVEL_ERROR, LOG_LEVEL_INFO, LOG_LEVEL_DEBUG, LOG_LEVEL_DEBUG_MORE)
#define WEB_LOG_LEVEL          LOG_LEVEL_INFO    // [WebLog] (LOG_LEVEL_NONE, LOG_LEVEL_ERROR, LOG_LEVEL_INFO, LOG_LEVEL_DEBUG, LOG_LEVEL_DEBUG_MORE)
#define MQTT_LOG_LEVEL         LOG_LEVEL_NONE    // [MqttLog] (LOG_LEVEL_NONE, LOG_LEVEL_ERROR, LOG_LEVEL_INFO, LOG_LEVEL_DEBUG, LOG_LEVEL_DEBUG_MORE)

// -- Ota -----------------------------------------
#ifdef ESP8266
#define OTA_URL                "http://ota.tasmota.com/tasmota/release/tasmota.bin.gz"  // [OtaUrl]
#endif  // ESP8266
#ifdef ESP32
#define OTA_URL                "http://ota.tasmota.com/tasmota32/release/tasmota32.bin"  // [OtaUrl]
#endif  // ESP32

// -- MQTT ----------------------------------------
#define MQTT_USE               true              // [SetOption3] Select default MQTT use (false = Off, true = On)

#define MQTT_HOST              ""                // [MqttHost]
#define MQTT_FINGERPRINT1      0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00  // [MqttFingerprint1] (auto-learn)
#define MQTT_FINGERPRINT2      0xDA,0x39,0xA3,0xEE,0x5E,0x6B,0x4B,0x0D,0x32,0x55,0xBF,0xEF,0x95,0x60,0x18,0x90,0xAF,0xD8,0x07,0x09  // [MqttFingerprint2] (invalid)
#define MQTT_PORT              1883              // [MqttPort] MQTT port (10123 on CloudMQTT)
#define MQTT_USER              "DVES_USER"       // [MqttUser] MQTT user
#define MQTT_PASS              "DVES_PASS"       // [MqttPassword] MQTT password

#define MQTT_BUTTON_RETAIN     false             // [ButtonRetain] Button may send retain flag (false = off, true = on)
#define MQTT_POWER_RETAIN      false             // [PowerRetain] Power status message may send retain flag (false = off, true = on)
#define MQTT_SWITCH_RETAIN     false             // [SwitchRetain] Switch may send retain flag (false = off, true = on)
#define MQTT_SENSOR_RETAIN     false             // [SensorRetain] Sensor may send retain flag (false = off, true = on)
#define MQTT_INFO_RETAIN       false             // [InfoRetain] Info may send retain flag (false = off, true = on)
#define MQTT_STATE_RETAIN      false             // [StateRetain] State may send retain flag (false = off, true = on)
#define MQTT_NO_HOLD_RETAIN    false             // [SetOption62] Disable retain flag on HOLD messages
#define MQTT_NO_RETAIN         false             // [SetOption104] No Retain - disable all MQTT retained messages, some brokers don't support it: AWS IoT, Losant

#define MQTT_STATUS_OFF        "OFF"             // [StateText1] Command or Status result when turned off (needs to be a string like "0" or "Off")
#define MQTT_STATUS_ON         "ON"              // [StateText2] Command or Status result when turned on (needs to be a string like "1" or "On")
#define MQTT_CMND_TOGGLE       "TOGGLE"          // [StateText3] Command to send when toggling (needs to be a string like "2" or "Toggle")
#define MQTT_CMND_HOLD         "HOLD"            // [StateText4] Command to send when button is kept down for over KEY_HOLD_TIME * 0.1 seconds (needs to be a string like "HOLD")

// -- MQTT topics ---------------------------------
  // Example "tasmota/bedroom/%topic%/%prefix%/" up to 80 characers
#define MQTT_FULLTOPIC         "%prefix%/%topic%/" // [FullTopic] Subscribe and Publish full topic name - Legacy topic

// %prefix% token options
#define SUB_PREFIX             "cmnd"            // [Prefix1] Tasmota devices subscribe to %prefix%/%topic% being SUB_PREFIX/MQTT_TOPIC and SUB_PREFIX/MQTT_GRPTOPIC
#define PUB_PREFIX             "stat"            // [Prefix2] Tasmota devices publish to %prefix%/%topic% being PUB_PREFIX/MQTT_TOPIC
#define PUB_PREFIX2            "tele"            // [Prefix3] Tasmota devices publish telemetry data to %prefix%/%topic% being PUB_PREFIX2/MQTT_TOPIC/UPTIME, POWER and TIME
                                                 //   May be named the same as PUB_PREFIX
// %topic% token options (also ButtonTopic and SwitchTopic)
#define MQTT_TOPIC             PROJECT "_%06X"   // [Topic] unique MQTT device topic including (part of) device MAC address
#define MQTT_GRPTOPIC          "switch"        // [GroupTopic] MQTT Group topic
#define MQTT_GROUPTOPIC_FORMAT false             // [SetOption75] GroupTopic replaces %topic% (false) or fixed topic cmnd/grouptopic (true)
#define MQTT_BUTTON_TOPIC      "0"               // [ButtonTopic] MQTT button topic, "0" = same as MQTT_TOPIC, set to 'PROJECT "_BTN_%06X"' for unique topic including device MAC address
#define MQTT_SWITCH_TOPIC      "0"               // [SwitchTopic] MQTT button topic, "0" = same as MQTT_TOPIC, set to 'PROJECT "_SW_%06X"' for unique topic including device MAC address
#define MQTT_CLIENT_ID         "DVES_%06X"       // [MqttClient] Also fall back topic using last 6 characters of MAC address or use "DVES_%12X" for complete MAC address

// -- MQTT - Telemetry ----------------------------
#define TELE_PERIOD            300               // [TelePeriod] Telemetry (0 = disable, 10 - 3600 seconds)
#define TELE_ON_POWER          false             // [SetOption59] send tele/STATE together with stat/RESULT (false = Disable, true = Enable)

// -- MQTT - Domoticz -----------------------------
#define DOMOTICZ_UPDATE_TIMER  0                 // [DomoticzUpdateTimer] Send relay status (0 = disable, 1 - 3600 seconds)

// -- MQTT - Home Assistant Discovery -------------
#define HOME_ASSISTANT_DISCOVERY_ENABLE   true  // [SetOption19] Home Assistant Discovery (false = Disable, true = Enable)
#define HASS_AS_LIGHT          false             // [SetOption30] Enforce HAss autodiscovery as light
//#define DEEPSLEEP_LWT_HA_DISCOVERY             // Enable LWT topic and its payloads for read-only sensors (Status sensor not included) and binary_sensors on HAss Discovery (Commented out: all read-only sensors and binary_sensors
                                                 // won't be shown as OFFLINE on Home Assistant when the device is DeepSleeping - NOTE: This is only for read-only sensors and binary_sensors, relays will be shown as OFFLINE)

// -- MQTT - Options ------------------------------
#define MQTT_RESULT_COMMAND    false             // [SetOption4]  Switch between MQTT RESULT or COMMAND
#define MQTT_LWT_MESSAGE       false             // [SetOption10] Switch between MQTT LWT OFFLINE or empty message
#define MQTT_POWER_FORMAT      false             // [SetOption26] Switch between POWER or POWER1 for single power devices
#define MQTT_APPEND_TIMEZONE   false             // [SetOption52] Append timezone to JSON time
#define MQTT_BUTTON_SWITCH_FORCE_LOCAL   false   // [SetOption61] Force local operation when button/switch topic is set (false = off, true = on)
#define MQTT_INDEX_SEPARATOR   false             // [SetOption64] Enable "_" instead of "-" as sensor index separator
#define MQTT_TUYA_RECEIVED     false             // [SetOption66] Enable TuyaMcuReceived messages over Mqtt
#define MQTT_TLS_ENABLED       false             // [SetOption103] Enable TLS mode (requires TLS version)

// -- HTTP ----------------------------------------
#define WEB_SERVER             2                 // [WebServer] Web server (0 = Off, 1 = Start as User, 2 = Start as Admin)
#define WEB_PASSWORD           ""                // [WebPassword] Web server Admin mode Password for WEB_USERNAME (empty string = Disable)
#define FRIENDLY_NAME          "Switch"         // [FriendlyName] Friendlyname up to 32 characters used by webpages and Alexa
#define EMULATION              EMUL_HUE         // [Emulation] Select Belkin WeMo (single relay/light) or Hue Bridge emulation (multi relay/light) (EMUL_NONE, EMUL_WEMO or EMUL_HUE)
#define EMULATION_HUE_1ST_GEN  false             // [Emulation] Force SetOption109 1 - if you only have Echo Dot 2nd gen devices
#define CORS_DOMAIN            ""                // [Cors] CORS Domain for preflight requests

// -- HTTP Options --------------------------------
#define GUI_SHOW_HOSTNAME      false             // [SetOption53] Show hostname and IP address in GUI main menu
.
.
.
.
.
.


I don't understand what is going wrong.

**Reply by barbudor** [reply]
Are you modifying `my_user_config.h` ?
You are not supposed to
You should make your changes in `user_config_override.h`
If you start modifying a key file of Tasmota it makes helping you more difficult because I can't see immediatly what you have changed

**Reply by zeebak** [reply]
Yes I am sorry about the _my_user_config.h_ file. 

But only the USER_MODULE, USER_TEMPLATE  are changed.
I haven't touched the rest of the code.

#define USER_TEMPLATE "{\"NAME\":\"mytemplate\",\"GPIO\":[1,32,1,256,257,258,1,1,33,259,34,35,1,0],\"FLAG\":0,\"BASE\":18}"

**Reply by barbudor** [reply]
I just recompile tastmota with WIFI_CONFIG_TOOL WIFI_MANAGER
Then I erased my NodeMCU and serial-flashed from clean
- Booted to AP
- Entered my wifi credentials from mobile
- Rebooted, connected to my Wifi

And that's all
No reboot to AP, no stuck in AP

I can't reproduce your issue

**Reply by zeebak** [reply]
Thank you very much for your time.
Then I think it might be an issue with the hardware.
I am using esp 12f module.

_VCC                 3.3v
CH_PD            3.3 via 10k resistor
GPIO15           gnd via 10k resistor
RST pin           3.3v via 10k_

It resets to factory default every 35-40 seconds

**Reply by barbudor** [reply]
No. This is standard recovery procedure because you have "Button 1" on a pin that is constantly held "pushed".
If you are using GPIO15 for button, it should be "Button_in" (inverted, no pull-up) and button connected between GPIO15 and VCC

See https://tasmota.github.io/docs/Buttons-and-Switches/#button
and https://tasmota.github.io/docs/Buttons-schematics/

See https://tasmota.github.io/docs/Device-Recovery/ for explanation on recovery

**Reply by zeebak** [reply]
I feel so dumb I overlooked Button1.
Changed the Button1, now everything works as expected. Thank you so much!

**Reply by zeebak** [reply]
I changed the Buttons to Switches. Another problem I am facing, when switch is closed (vcc connected), the board fails to boot.
4 relays and 4 switches, Can you help me configure how would I connect these

**Reply by barbudor** [reply]
GPIO15 ???
Once again you don't fully read the docs

For correct boot, GPIO0 and GPIO2 MUST be VCC and GPIO15 MUST Be GROUND

You may connect push buttons to those GPIO as far are you are sure they are nots pressed at boot.
But do not connect switches.

**Reply by zeebak** [reply]
I have 4 relays connected that I want to control with buttons/switches. Using Switches would be easier for since I have many spare ones.
Out of the usable GPIO00, 01, 02, 03, 04, 05, 12, 13, 14, 15, 16
I can't use GPIO0, 01, 02, 03, 16 for the relays since they are high at boot and flickers the relays at boot.
So I chose 04, 05, 12, 13 for my relays and out of the rest, 4 pins I need for switches.

**Reply by barbudor** [reply]
Use a ESP32, more pins, less problems

**Reply by sfromis** [reply]
You can configure analog (gpio17) as an extra digital input button.

**Reply by zeebak** [reply]
@sfromis  Yes, but I was hoping to use Switches instead of Buttons.

**Reply by zeebak** [reply]
> Use a ESP32, more pins, less problems

I think I found a solution, I used GPIO 04, 05, 12, 13 (safe pins) for the switches
and 03, 14, 16 with 'Low triggered' relay and 15 with a 'High triggered' relay.
